Update conda packages to clang 11

cticenhour opened the pull request:
Updates moose-mpich conda recipe to use clang_osx-64 11.0.0 build hb91bd55_5 for new SDK support and updates related dependencies as a result of that change. 

Still to do:
- [x] Percolate changes to moose-mpich to other packages
- [x] Fixup new warnings in libMesh (see libmesh/libmesh#2812), gtest, and MOOSE
- [x] ~~Update any needed Linux dependencies~~ EDIT: none needed

Tagging @milljm 

Refs #16291

cticenhour committed:
Update moose-mpich to clang_osx-64 11.0.0 build hb91bd55 and update related dependencies (#16291)

cticenhour committed:
Fixup range-loop-construct warning (#16291)

cticenhour committed:
Ignore gtest deprecated copy warning for clang 10+ (#16291)

cticenhour committed:
Fixup deprecated copy warning for clang 10+ in MultiIndex (#16291)

cticenhour commented:
@lindsayad A libmesh update will need to be performed for this PR after libmesh/libmesh#2812 is finished and merged. Would you rather do that in a separate PR or as part of this?

lindsayad commented on framework/contrib/gtest/gtest/gtest.h:
```diff
@@ -51,6 +51,15 @@
 #ifndef GTEST_INCLUDE_GTEST_GTEST_H_
 #define GTEST_INCLUDE_GTEST_GTEST_H_
 
+// This warning is a known issue since 2019 (https://github.com/google/googletest/issues/2224)
+// Google has yet to fix it, so we will direct clang > 9 to ignore the warnings.
+#ifdef __clang__
+#if (__clang_major__ > 9)
+// This was introduced in clang 10
+#pragma clang diagnostic ignored "-Wdeprecated-copy"
+#endif // clang > 9
+#endif
+
```
Will this lead to ignoring bad copies in MOOSE unit test code? I'm guessing so

cticenhour replied:
I didn't consider that - this has been a longstanding issue in gtest, so I am unsure how to handle it short of diving into that code and fixing up our own copy.

lindsayad replied:
It depends on when the warnings happen. There are so many macros with gtest that *a priori*, without actually incurring the warnings during a build, it's impossible to know what macros spawn warnings. For instance, if this
```c++

TEST(ADFparserTest, JITCompile)
{
```
spawns a warning it would be obnoxious to put this diagnostic ignored pragma around every one of the macro instances.

lindsayad replied:
I don't know. What you have is probably ok. We should just remove this as soon as gtest fixes their shizz

cticenhour replied:
I could always temporarily remove and see what happens when I actually run the test suite. Do you know if we have `-Werror` turned on for that?

cticenhour replied:
I'm actually a little tempted to fix this up myself - we're already a decent amount behind HEAD and it looks like gtest doesn't accept PRs. I'll ponder over the holiday :D

cticenhour replied:
Well perhaps not.....I thought we had forked gtest, but it seems as though the last change to this file was a superficial one.

lindsayad commented on framework/src/postprocessors/NumRelationshipManagers.C:
```diff
@@ -46,19 +46,19 @@ NumRelationshipManagers::getValue()
   if (_rm_type == "GEOMETRIC")
   {
-    for (const auto rm : rms)
+    for (const auto & rm : rms)
```
`rm` is a pointer, so let's instead do `&` -> `* const`

cticenhour replied:
Oops - I got lazy and followed the compiler's suggestion. 😅 Thanks!

cticenhour replied:
Actually, clang didn't like that extra `const`, since the type ends up as `const std::shared_ptr<RelationshipManager>`, so I'll stick with just a `*`.

cticenhour replied:
Clang isn't happy with either `*` or `* const`:

```
/Users/icenct/projects/moose/framework/src/postprocessors/NumRelationshipManagers.C:49:23: error: variable 'rm' with type 'const auto *' has incompatible initializer of type 'const std::__1::shared_ptr<RelationshipManager>'
    for (const auto * rm : rms)
                      ^  ~
```

Shouldn't this declared variable be a reference to that shared pointer?

lindsayad replied:
Then I've led you astray. Since it's a `shared_ptr`, then the `&`s are good

lindsayad commented on framework/src/postprocessors/NumRelationshipManagers.C:
```diff
@@ -46,19 +46,19 @@ NumRelationshipManagers::getValue()
   else if (_rm_type == "ALGEBRAIC")
   {
-    for (const auto rm : rms)
+    for (const auto & rm : rms)
```
`&` -> `* const`

lindsayad commented on framework/src/postprocessors/NumRelationshipManagers.C:
```diff
@@ -46,19 +46,19 @@ NumRelationshipManagers::getValue()
   else if (_rm_type == "COUPLING")
   {
-    for (const auto rm : rms)
+    for (const auto & rm : rms)
```
`&` -> `* const`

lindsayad reviewed (dismissed):
I love these compiler updates. They elucidate bad code

lindsayad commented:
> @lindsayad A libmesh update will need to be performed for this PR after [libMesh/libmesh#2812](https://github.com/libMesh/libmesh/pull/2812) is finished and merged. Would you rather do that in a separate PR or as part of this?

Probably separate

lindsayad reviewed (commented):
`shared_ptr ` instead of raw pointer...my concerns are invalid :smile:

lindsayad commented:
Omg I have to explicitly dismiss my requested changes if I don't explicitly approve :-/

cticenhour commented:
FYI - the changes in libMesh have been merged into master.

cticenhour committed:
Update remaining conda packages modified by moose-mpich changes (#16291)

cticenhour commented:
@milljm Did you want to do any updating with the Linux dependencies here? Or just leave gcc etc as-is since this is a clang-centric change due to Xcode 12?

milljm commented:
The recipe changes look fine. But I would hold off until we fix firewall rules on civet (anaconda channel will never get updated with these new packages).

milljm commented:
I am seeing a lot of warning while using LLVM 11 (in Spack). Not sure how things look on Conda.

cticenhour commented:
I added some warnings to libMesh `ignore_warnings.h` that were popping up in the moose build for clang 11 and fixed up some others in this PR. We will still see some warnings related to libmesh contribs that will hang around for now, after my conversations with Roy in libmesh/libmesh#2812. So the next libMesh update should take care of some of what you're seeing, and this PR should take care of much of the rest.

cticenhour commented:
That is, if the warnings we're speaking of are the same ones.

cticenhour commented:
@lindsayad When would you like to do a libmesh update? I'd like to get testing started on this, and I know that it'll fail currently with some of our recipes that use `-Werror`.

milljm commented:
We can't perform one yet. Not until our Firewall rules get fixed (See how master is unable to update any Anaconda packages: https://civet.inl.gov/branch/4/)

cticenhour commented:
Right, we talked about that yesterday. Not looking for a merge due to that issue, but I was under the impression that modifications to conda recipes got PR tested. Was I wrong?

cticenhour commented:
Or are the stacks not built for the first time until next?

milljm commented:
You are correct, you can create PR's all you like. I am just stating: don't merge that into next until we fix firewall issues.

lindsayad commented:
There are a few things I would like to see resolved before a libmesh update: libmesh/libmesh#2805 (@fdkong) and libmesh/libmesh#2749 (@loganharbour)

loganharbour commented:
Thanks @lindsayad. libMesh/libmesh#2749 should be good this afternoon.

cticenhour commented:
Sounds good - thanks for the updates everybody.

fdkong commented:
> There are a few things I would like to see resolved before a libmesh update: [libMesh/libmesh#2805](https://github.com/libMesh/libmesh/pull/2805) (@fdkong) and [libMesh/libmesh#2749](https://github.com/libMesh/libmesh/pull/2749) (@loganharbour)

My libmesh PR is closed because I do not like the shared-pointer idea, and you guys free to go ahead for libmesh update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/644040/) on aa2e927 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16295/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16295/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4321b5c1d6c6256aa4902702a589999df8854977`

Commit:
Remove src requirement in moose-compilers

No longer need a src of any kind when only installing other packages.

cticenhour committed:
Fixup warnings in phase field and tensor mechanics modules for clang 11 (#16291)

cticenhour committed:
Address comment - limit gtest ignore warnings to gtest (#16291)

cticenhour commented on framework/contrib/gtest/gtest/gtest.h:
```diff
@@ -21198,4 +21199,9 @@ inline int RUN_ALL_TESTS() {
   return ::testing::UnitTest::GetInstance()->Run();
 }
 
+// Restore warnings to other classes (see top of file for info on ignored warnings)
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif
```
@lindsayad I think this and Line 57 addresses your previous comment about this ignore impacting other unit test objects. The "push/pop" pragmas are the same used in the libmesh `ignore_warnings.h` and `restore_warnings.h` to isolate what is ignored.

cticenhour commented:
@lindsayad I have rebased on the submodule updates you and Fande merged today, yet I am still seeing the `/usr/bin/bash` issue. Packages are being built on CentOS 8 and tested on Ubuntu 16.04. See https://civet.inl.gov/job/646099/

lindsayad commented:
Ok yea I incorrectly identified the issue. Here’s what I think is going on: is this the first time we’ve built mpich on meowbuild @milljm ? I bet it is and mpich has /usr/bin stuff built into it now. I don’t have  any workarounds for that one. At the point it’s in the mpi compilers we’re pretty well done for

milljm commented:
This is the first time we built the compilers on CentOS-7.7. Looks like we are going back to building the package on CentOS 7.6.

moosebuild commented:
All [jobs](https://civet.inl.gov/event/61214/) on 6788266 : invalidated by @milljm

Created CentOS 7.6 Container for Create Linux Stack

cticenhour commented:
I haven't taken a look at the conda build and configuration scripts yet for this case, but are there edits we can make to get the mpich compiler package to look in the right place? Or would this be something deeper in this particular version of mpich?

lindsayad commented:
This isn't something you can fix on the MPI side. It's all about PATH ordering on the configure/build system. If `/usr/bin` is ahead of `/bin` in PATH and `/usr/bin` has all the core utilities, then configure is going to pick up the `/usr/bin` utilities. And whereas `/bin` core utilities are guaranteed to exist on all systems (because a machine has to be able to work regardless of users), `/usr/bin` core utilities are not guaranteed to work on all systems

cticenhour commented:
Right, I am aware of the distinction now between systems for this issue and that it is a configure/build issue - I was just wondering if there was some way we can use the configure or build scripts to standardize a bit for us without causing more headaches down the line. Like defining `PATH` more consistently in our build script to override system differences. I could see that one being dicey though.

milljm commented:
I know this is confusing, but I think this PR fails to update moose-libmesh properly. Not that _this_ is the problem with building libmesh in the rest of the targets (I am looking into that). But this is what I end up with when I search for moose-libmesh after adding this PR's channel:

```
Loading channels: done
# Name                       Version           Build  Channel             
moose-libmesh             2020.06.02         build_0  idaholab            
moose-libmesh             2020.06.03         build_0  idaholab            
moose-libmesh             2020.06.16         build_0  idaholab            
moose-libmesh             2020.06.16         build_1  idaholab            
moose-libmesh             2020.07.23         build_0  idaholab            
moose-libmesh             2020.08.06         build_0  idaholab            
moose-libmesh             2020.10.05         build_0  idaholab            
moose-libmesh             2020.11.04         build_0  idaholab            
moose-libmesh             2020.11.04         build_1  idaholab            
moose-libmesh             2020.12.04         build_1  idaholab            
moose-libmesh             2020.12.10         build_0  moose/pr-16295      
moose-libmesh             2020.12.10         build_0  idaholab 
```

As you can see, there will be a conflict. Well, not really a conflict, but this PR's version of libMesh will overwrite the current one that is out there _not_ built with LLVM 11.

I don't have a good way of always keeping the last-PR-to-make-into-next, knowing when to bump the version number. It was either Alex's or Fande's libMesh update that is now 2020.12.10 build_0.

cticenhour commented:
@milljm I imagine there's not a really good way of keeping that consistent beyond being vigilant....I'll go ahead and increment the build number for the YAML file so we can avoid the overwrite in this case.

cticenhour committed:
Increment libMesh package build number to 1 (#16291)

moosebuild commented:
Job [External App tests (Conda libMesh)](https://civet.inl.gov/job/646112/) on 6788266 : invalidated by @loganharbour

Machine died

moosebuild commented:
Job [Python 3.6 Tests](https://civet.inl.gov/job/646099/) on 6788266 : invalidated by @loganharbour

Machine died

fdkong commented:
> Right, I am aware of the distinction now between systems for this issue and that it is a configure/build issue - I was just wondering if there was some way we can use the configure or build scripts to standardize a bit for us without causing more headaches down the line. Like defining `PATH` more consistently in our build script to override system differences. I could see that one being dicey though.

Nobody ever try to set an environment  like "SHELL=bin/bash"? Before we build compilers

milljm commented:
I thought about adjusting path when ./build.sh is run. But I am not sure what that might do to Conda before it hands control over to `build.sh`.

I suppose all we can do is try it?

```bash
export PATH=/usr/bin:$PATH
```
at the top of each `build.sh`.

lindsayad commented:
I'd be interested to give that a whirl

Commit:
Adjust PATH while building

Attempt to discover all tools in /bin before /usr/bin. Some operating systems
(Ubuntu 16.x and 18.x) are missing tools in /bin, when they are instead in
/usr/bin. Unfortunately, some packages record where they discover these
binaries and are thus considered 'missing' on another OS after insallation.

cticenhour commented:
Seems to build without a hitch, but doesn't make a difference at testing time....https://civet.inl.gov/job/646684/

cticenhour commented:
Can conda packages edit `PATH` when you activate the environment? Perhaps that's the way.

cticenhour commented:
Interestingly, this works on another Ubuntu 16 instance https://civet.inl.gov/job/646693/

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/646689/) on 1f5d0f1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16295/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Boost enabled, Pthreads only](https://civet.inl.gov/job/646684/) on 1f5d0f1 : invalidated by @fdkong

Should work

milljm commented:
I am not quite understanding what is slowing down the non-meowbuild machines when dealing with Conda. +1hr seems common now. Even on our Trashcan machines, it seems to be taking longer.

cticenhour commented:
Just judging from the intermittent failures I saw late yesterday, it seems like the cleaning step we have in place just isn't enough sometimes. One invalidation caught an "unrecoverable conda issue" and did a hard reset successfully, but then the other performed a standard cleaning and proceeded.

milljm commented:
You wouldn't have a link to both of these events would you?

cticenhour commented:
Here you go: 

- This one failed, Fande invalidated late last night, and it passed after a reinitialize and cleaning. Unfortunately I do not have a record of the previous results before the invalidation: https://civet.inl.gov/job/646684/
- This one failed, I invalidated it yesterday afternoon, and it passed after experiencing a conda issue and performing a reset during the Initialize step: https://civet.inl.gov/job/646692/
- This passed on the first go, but did require the same reset during Initialize:  https://civet.inl.gov/job/646693/

milljm commented:
I'll add a print statement to the diff command that detects a 'tainted' environment. See if we can't spot why this happens all the time.

lindsayad commented on framework/contrib/gtest/gtest/gtest.h:
```diff
@@ -51,6 +51,16 @@
 #ifndef GTEST_INCLUDE_GTEST_GTEST_H_
 #define GTEST_INCLUDE_GTEST_GTEST_H_
 
+// This warning is a known issue since 2019 (https://github.com/google/googletest/issues/2224)
+// Google has yet to fix it, so we will direct clang > 9 to ignore the warnings.
+#ifdef __clang__
+#pragma clang diagnostic push
+#if (__clang_major__ > 9)
+// This was introduced in clang 10
+#pragma clang diagnostic ignored "-Wdeprecated-copy"
+#endif // clang > 9
+#endif
```
This is exactly what I was hoping to see. Thanks!

cticenhour replied:
Thanks for prompting the change! I didn't know about these pragmas until I dug into those libmesh headers a bit more! :D

lindsayad reviewed (approved):
LGTM!

cticenhour commented:
FYI - holding merge for now until I have a go from Dev-Ops folks.

milljm reviewed (approved):
👍 
LLVM 11! Nice!
